SSO Cluster Discussion (#718)
BCDevOps/platform-services: apps/sso/docs/cluster-sso.md

- Openshift usernames map directly to a github username.

## Option 1: Use SSO IDIR for Authentication/Authorization

Ditch the usage of the Github OAuth App. Users will authenticate with the IDIRs via SSO. 

There are a few pros/cons with this approach.

### Pros
- authentication is aligned with Gov's standards of how users access gov services

### Cons
- IDIR does not allow for 2fa
- Creates (as of now) a blocker for non government teams to access a cluster. They would need to procure IDIRs in order to authentication. This process
requires manual intervention (iStore orders and HR processes). 
- Unable to use Github as an IDP. Github oidc does not provide a way to scope users to a particular Github Org. All github users would be authenticated.
Currently there is no way to provide a logic layer to authorize users (a RH ticket has been opened by Shelly inquiring on this process).

## Option 2: Use Github OAuth (same implementation as 3.11)

Business as usual. We have a good POC here. It has been working well for two years. A user is added to the BCDevops Github org. Openshift leverages a Github OAuth app to authenticate users against that organization.

> 

Reviewer: ```suggestion

We could also add a second OAuth provider (SSO) to provide IDIR auth without leveraging the SSO-GitHub provider.  This would allow both IDIR as well as a scoped GitHub as authentication providers.
```

## Pros

> - 2fa enabled

Reviewer: ```suggestion
- 2fa enabled (for GitHub identities)
```

- We can easily establish automation via the project registry to add users to the BCDevopsOrg for access to the cluster
- You can create Github Accounts quickly

> 

Reviewer: ```suggestion
- Addition of IDIR as an optional provider

```

## Cons
- There is a cost associated with Github Org Memebership
- __Not aligned__ with Gov's standards of how users access gov services

> 

Reviewer: ```suggestion
- Creates potentially non-uniform access within the cluster where userID's could be of 2 formats: 
eg: jefkel and jefkel@idir  - the github access would not contain the @provider syntax unless we are able to create a filtered provider within SSO.

```

Author: ugh! I forgot to add this in. Good point!

- Authentication source (IDIR|GITHUB) are identified within openshift to allow easier RBAC management. (eg: easy to separate the auth provider in the username for policy clarity)
### Con:
- Unable to filter/limit Github users by Organization (and force 2FA via organizational requirement)
- No 2FA for IDIR access (combined with above, means we cannot force 2FA for cluster authentication)

## Option 2: Replace all Auth with SSO (both IDIR and GitHub) (with tech preview features)

### Pro:
- Same as Option 1.
### Con:
- Additional flow development time needed to create custom RBAC/filter flows without current support (more investigation needed to determine approach to use)

## Option 3: Add multiple auth providers to cluster and limit by oidc configuration
### Pro:
- Allows existing approach for Github access (2FA enforcement, limited org scope, etc)
- Adds second provider for IDIR authentication flow
### Con:
- OpenShift RBAC clarity - Usernames in openshift are not as clearly delineated as to their authentication provider source (github users: *jefkel*, idir users: *jefkel@idir*) (could create developer team RBAC management headaches)

## Decision Point

> Based on the above factors, the team has decided to go with __Option 3__. Security was the main motivator behind this decision. There is ___no guarantee__ that SSO Tech preview features will released in production. In addition the very definition of __tech preview__ means that these features are not available for support by Redhat. 

Reviewer: ```suggestion
Based on the above factors, the team has decided to go with __Option 2__.  With the ability to leverage a custom post-login flow using existing (GA) features, we are able to avoid a dependency on a tech-preview feature, and still provide both the filtered restrictions for authentication, and the consistent/predictable user naming conventions for in-cluster RBAC. 
```

> 

Reviewer: ```suggestion

A separate SSO Realm with an IDIR IDP configured will be used by the clusters.  This realm will also be configured with SSO Multi Factor Authentication (MFA).  The SSO MFA will run for this realm only, and will provide interim MFA for the IDIR authentication service.  Once the new 2FA features for IDIR have been implemented across the IDIR Identity provider, this feature will be turned off.
```
